mlflow/store/tracking/sqlalchemy_store.py

```diff
@@ -338,6 +339,65 @@ def list_experiments(
             view_type=view_type, max_results=max_results, page_token=page_token, eager=True
         )
 
+    def _search_experiments(
+        self,
+        view_type,
+        max_results,
+        filter_string,
+        order_by,
+        page_token,
+    ):
+        def compute_next_token(current_size):
+            next_token = None
+            if max_results == current_size:
+                final_offset = offset + max_results
+                next_token = SearchExperimentsUtils.create_page_token(final_offset)
+
+            return next_token
+
+        if max_results and max_results > SEARCH_MAX_RESULTS_THRESHOLD:
+            raise MlflowException(
+                "Invalid value for request parameter max_results. It must be at "
+                "most {}, but got value {}".format(SEARCH_MAX_RESULTS_THRESHOLD, max_results),
+                INVALID_PARAMETER_VALUE,
+            )
```

Comment:
This check seems to imply that `max_results` can be `None`; in that case, what happens when we pass it to `limit()` on line 380?

The docstring for `AbstractStore.search_experiments()` indicates that, if max results is unspecified, the SqlAlchemyStore will return all results. If that's true, it's a bit strange that we throw an exception if `max_results` is specified and is too big, since we clearly support returning all available experiments by leaving `max_results` unspecified.

Shall we change the documentation of the AbstractStore method?

Reply:
Looks like the MlflowClient also has the same message; shall we change it?

Reply:
> what happens when we pass it to limit() on line 380?

`limit(None)` would fetch all experiments ([related SO question](https://stackoverflow.com/questions/12836194/how-do-i-remove-a-limit-that-is-already-applied-to-a-query-object))

> it's a bit strange that we throw an exception if max_results is specified and is too big, since we clearly support returning all available experiments by leaving max_results unspecified.

Yeah it's strange. I'll update the MlflowClient and AbstractStore docstrings.

Reply:
Can we

- Set the default value of `max_results` to `SEARCH_MAX_RESULTS_DEFAULT`
- Update the validation like this:

```python
        if not isinstance(max_results, int) and max_results < 1:
            raise MlflowException(
                "max_results must be a positive integer, but got {}".format(max_results),
                INVALID_PARAMETER_VALUE,
            )
       
        if max_results > SEARCH_MAX_RESULTS_THRESHOLD:
            raise MlflowException(
                "Invalid value for request parameter max_results. It must be at"
                " most {}, but got value {}".format(SEARCH_MAX_RESULTS_THRESHOLD, max_results),
                INVALID_PARAMETER_VALUE,
            )
```

- Update the doc like this:

```
        :param max_results: The maximum number of experiments desired. For the REST backend,
                            the server may apply its own limit.
```

?

Reply:
Sounds good!